core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -63,6 +63,7 @@ class ControllerContext {
   val topicsWithDeletionStarted = mutable.Set.empty[String]
   val topicsIneligibleForDeletion = mutable.Set.empty[String]
 
+  @volatile var livePreferredControllerIds: Set[Int] = Set.empty
```

Comment:
If I see it right, this variable will only be accessed and updated by the controller thread. Why do we need the `volatile` annotation here?

Reply:
Because the variable is also accessed through Kafka API request handler thread via AdminManager, so it is volatile.